src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+
+      LMS_Tree_Node_Idx q() const {
+         BOTAN_STATE_CHECK(is_leaf());
+         return LMS_Tree_Node_Idx(m_r - (1 << m_h.get()));
+      }
+
+   private:
+      LMS_TreeLayerIndex m_h;
+      uint32_t m_r;
+};
+
+std::function<void(
+   StrongSpan<LMS_Tree_Node>, const TreeAddress&, StrongSpan<const LMS_Tree_Node>, StrongSpan<const LMS_Tree_Node>)>
+get_hash_pair_func_for_identifier(const LMS_Params& lms_params, LMS_Identifier identifier) {
+   // hash object must be shared, otherwise std::function would not be copyable, which is not allowed
+   std::shared_ptr<HashFunction> hash = HashFunction::create_or_throw(lms_params.hash_name());
+   return [hash, I = std::move(identifier)](StrongSpan<LMS_Tree_Node> out,
+                                            const TreeAddress& address,
+                                            StrongSpan<const LMS_Tree_Node> left,
+                                            StrongSpan<const LMS_Tree_Node> right) {
+      auto lms_address = dynamic_cast<const TreeAddress&>(address);
+
+      hash->update(I);
+      hash->update_be(lms_address.r());
+      hash->update_be(D_INTR);
+      hash->update(left);
+      hash->update(right);
+      hash->final(out);
+   };
+}
+
+void lms_gen_leaf(StrongSpan<LMS_Tree_Node> out,
+                  const LMOTS_Public_Key& lmots_pk,
+                  const TreeAddress& tree_address,
+                  HashFunction& hash) {
+   hash.update(lmots_pk.identifier());
+   hash.update_be(tree_address.r());
+   hash.update_be(D_LEAF);
+   hash.update(lmots_pk.K());
+   hash.final(out);
+}
+
+std::function<void(StrongSpan<LMS_Tree_Node> out, const TreeAddress& address)> lms_gen_leaf_func(
+   const LMS_PrivateKey& lms_sk) {
+   // hash object must be shared, otherwise std::function would not be copyable, which is not allowed
+   std::shared_ptr<HashFunction> hash = HashFunction::create_or_throw(lms_sk.lms_params().hash_name());
+   return [lms_sk, hash](StrongSpan<LMS_Tree_Node> out, const TreeAddress& tree_address) {
+      auto lmots_sk = LMOTS_Private_Key(lms_sk.lmots_params(), lms_sk.identifier(), tree_address.q(), lms_sk.seed());
+      auto lmots_pk = LMOTS_Public_Key(lmots_sk);
+      lms_gen_leaf(out, lmots_pk, tree_address, *hash);
+   };
+}
+
+void lms_treehash(StrongSpan<LMS_Tree_Node> out_root,
+                  std::optional<StrongSpan<LMS_AuthenticationPath>> out_auth_path,
+                  std::optional<LMS_Tree_Node_Idx> leaf_idx,
+                  const LMS_PrivateKey& lms_sk) {
+   auto hash_pair_func = get_hash_pair_func_for_identifier(lms_sk.lms_params(), lms_sk.identifier());
+   auto gen_leaf = lms_gen_leaf_func(lms_sk);
+   TreeAddress lms_tree_address(lms_sk.lms_params().h());
+
+   treehash(out_root,
+            out_auth_path,
+            leaf_idx,
+            lms_sk.lms_params().m(),
+            LMS_TreeLayerIndex(lms_sk.lms_params().h()),
+            0,
+            hash_pair_func,
+            gen_leaf,
```

Comment:
As stated above: When these are lambdas that have a closured `unique_ptr<>` they won't be able to be copied.

```suggestion
            std::move(hash_pair_func),
            std::move(gen_leaf),
```